tensorboard/plugins/scalar/scalars_plugin_test.py

```diff
@@ -108,15 +110,32 @@ def generate_run_to_db(self, experiment_name, run_name):
           sess.run(summaries, feed_dict=feed_dict)
         sess.run(flush_op)
 
-  def testRoutesProvided(self):
-    """Tests that the plugin offers the correct routes."""
-    self.set_up_with_runs([self._RUN_WITH_SCALARS])
-    routes = self.plugin.get_plugin_apps()
-    self.assertIsInstance(routes['/scalars'], collections.Callable)
-    self.assertIsInstance(routes['/tags'], collections.Callable)
-
-  def generate_run(self, run_name):
-    subdir = os.path.join(self.logdir, run_name)
+  def with_runs(run_names):
+    """Run a test with a bare multiplexer and with a `data_provider`."""
+    def decorator(fn):
+      @functools.wraps(fn)
+      def wrapper(self, *args, **kwargs):
+        (logdir, multiplexer) = self.load_runs(run_names)
+        with self.subTest('bare multiplexer'):
```

Comment:
FWIW, was there a reason you prefer this approach over defining a mixin class with a set of common test methods, and then separate test cases that define different objects under test?  That's the approach I've seen more often (and done more often myself).

Reply:
It’s easier for me to reason about how decorators compose than about how
method resolution order works. With subclassing, you also have you think
about where any state should live. No strong preference.